plugins/mac-virtualcam/src/obs-plugin/plugin-main.mm

```diff
@@ -0,0 +1,201 @@
+{
+	(void)type_data;
+	return obs_module_text("macOS Virtual Webcam");
+}
+
+// This is a dummy pointer so we have something to return from virtualcam_output_create
+static void *data = &data;
+
+static void *virtualcam_output_create(obs_data_t *settings,
+				      obs_output_t *output)
+{
+	blog(LOG_DEBUG, "output_create");
+	sMachServer = [[MachServer alloc] init];
+	return data;
+}
+
+static void virtualcam_output_destroy(void *data)
+{
+	blog(LOG_DEBUG, "output_destroy");
+	sMachServer = nil;
+}
+
+static bool virtualcam_output_start(void *data)
+{
+	blog(LOG_DEBUG, "output_start");
+
+	[sMachServer run];
+
+	obs_get_video_info(&videoInfo);
+
+	struct video_scale_info conversion = {};
+	conversion.format = VIDEO_FORMAT_UYVY;
+	conversion.width = videoInfo.output_width;
+	conversion.height = videoInfo.output_height;
+	obs_output_set_video_conversion(output, &conversion);
+	if (!obs_output_begin_data_capture(output, 0)) {
+		return false;
+	}
+
+	return true;
+}
+
+static void virtualcam_output_stop(void *data, uint64_t ts)
+{
+	blog(LOG_DEBUG, "output_stop");
+	obs_output_end_data_capture(output);
+	[sMachServer stop];
+}
+
+static void virtualcam_output_raw_video(void *data, struct video_data *frame)
+{
+	uint8_t *outData = frame->data[0];
+	if (frame->linesize[0] != (videoInfo.output_width * 2)) {
+		blog(LOG_ERROR,
+		     "unexpected frame->linesize (expected:%d actual:%d)",
+		     (videoInfo.output_width * 2), frame->linesize[0]);
+	}
+
+	CGFloat width = videoInfo.output_width;
+	CGFloat height = videoInfo.output_height;
+
+	[sMachServer sendFrameWithSize:NSMakeSize(width, height)
+			     timestamp:frame->timestamp
+			  fpsNumerator:videoInfo.fps_num
+			fpsDenominator:videoInfo.fps_den
+			    frameBytes:outData];
+}
+
+static bool GetDALPluginStatus()
+{
+	NSFileManager *fileManager = [NSFileManager defaultManager];
+
+	NSString *pathForFile =
+		@"/Library/CoreMediaIO/Plug-Ins/DAL/obs-mac-virtualcam.plugin";
+
+	return [fileManager fileExistsAtPath:pathForFile];
+}
+
+static void CopyDALPlugin()
+{
+	NSString *sourcePath;
+	NSRunningApplication *app = [NSRunningApplication currentApplication];
+
+	if ([app bundleIdentifier] != nil) {
+		NSURL *bundleURL = [app bundleURL];
+		NSString *pluginPath =
+			@"Contents/Resources/data/obs-mac-virtualcam.plugin";
+
+		NSURL *pluginUrl =
+			[bundleURL URLByAppendingPathComponent:pluginPath];
+		sourcePath = [pluginUrl path];
+	} else {
+		sourcePath = [[[[app executableURL]
+			URLByAppendingPathComponent:
+				@"../data/obs-mac-virtualcam.plugin"] path]
+			stringByReplacingOccurrencesOfString:@"obs/"
+						  withString:@""];
+	}
+
+	NSFileManager *fileManager = [NSFileManager defaultManager];
+
+	if ([fileManager fileExistsAtPath:sourcePath]) {
+		NSString *destinationPath =
+			@"/Library/CoreMediaIO/Plug-Ins/DAL/obs-mac-virtualcam.plugin";
+		NSString *copyCmd = [NSString
+			stringWithFormat:
+				@"do shell script \"cp -R '%@' '%@'\" with administrator privileges",
```

Comment:
This is neat

Reply:
Yeah originally I tried popping up a directory browsing prompt to get write permissions for the Library folder, but that doesn't work if OBS doesn't run as a sandboxed app and requires to run as an App bundle to pick up the necessary entitlements and all. In the end running an AppleScript was a bit dirty, but did the trick.